src/box/vinyl.c

```diff
@@ -4342,14 +4363,22 @@ vinyl_space_build_index(struct space *src_space, struct index *new_index,
 			if (rc != 0)
 				break;
 		}
+		ERROR_INJECT_DOUBLE(ERRINJ_BUILD_INDEX_TIMEOUT, inj->dparam > 0,
+				    thread_sleep(inj->dparam));
 		/*
 		 * Read iterator yields only when it reads runs.
 		 * Yield periodically in order not to stall the
 		 * tx thread in case there are a lot of tuples in
 		 * mems or cache.
 		 */
-		if (++loops % VY_YIELD_LOOPS == 0)
+		if (++loops % VY_YIELD_LOOPS == 0) {
```

Comment:
Please do it for the memtx engine and move it to a separate patch.

Reply:
Done.

<details>
<summary>patch</summary>

```diff
commit b464a4368d06139ceb2cec20e170df7708c0cb19
Author: Nikolay Shirokovskiy <nshirokovskiy@tarantool.org>
Date:   Tue Feb 13 12:34:41 2024 +0300

    engine: make secondary index build cancellable
    
    So that graceful shutdown during heavy secondary index build is possible.
    
    Part of #8423
    
    NO_CHANGELOG=internal
    NO_DOC=internal

diff --git a/src/box/memtx_space.c b/src/box/memtx_space.c
index 12a9d14cf..eea29eb75 100644
--- a/src/box/memtx_space.c
+++ b/src/box/memtx_space.c
@@ -1347,6 +1347,10 @@ memtx_space_build_index(struct space *src_space, struct index *new_index,
 		 */
 		if (!can_yield)
 			continue;
+
+		ERROR_INJECT_DOUBLE(ERRINJ_BUILD_INDEX_TIMEOUT, inj->dparam > 0,
+				    thread_sleep(inj->dparam));
+
 		/*
 		 * Remember the latest inserted tuple to
 		 * avoid processing yet to be added tuples
@@ -1363,6 +1367,11 @@ memtx_space_build_index(struct space *src_space, struct index *new_index,
 		 */
 		ERROR_INJECT_YIELD(ERRINJ_BUILD_INDEX_DELAY);
 		tuple_unref(state.cursor);
+		if (fiber_is_cancelled()) {
+			diag_set(FiberIsCancelled);
+			rc = -1;
+			break;
+		}
 		/*
 		 * The on_replace trigger may have failed
 		 * during the yield.
diff --git a/src/box/vinyl.c b/src/box/vinyl.c
index 165f8709f..44dc0160c 100644
--- a/src/box/vinyl.c
+++ b/src/box/vinyl.c
@@ -4372,14 +4372,22 @@ vinyl_space_build_index(struct space *src_space, struct index *new_index,
 			if (rc != 0)
 				break;
 		}
+		ERROR_INJECT_DOUBLE(ERRINJ_BUILD_INDEX_TIMEOUT, inj->dparam > 0,
+				    thread_sleep(inj->dparam));
 		/*
 		 * Read iterator yields only when it reads runs.
 		 * Yield periodically in order not to stall the
 		 * tx thread in case there are a lot of tuples in
 		 * mems or cache.
 		 */
-		if (++loops % VY_YIELD_LOOPS == 0)
+		if (++loops % VY_YIELD_LOOPS == 0) {
 			fiber_sleep(0);
+			if (fiber_is_cancelled()) {
+				diag_set(FiberIsCancelled);
+				rc = -1;
+				break;
+			}
+		}
 		if (ctx.is_failed) {
 			diag_move(&ctx.diag, diag_get());
 			rc = -1;
diff --git a/test/app-luatest/shutdown_test.lua b/test/app-luatest/shutdown_test.lua
index a7a317875..323145941 100644
--- a/test/app-luatest/shutdown_test.lua
+++ b/test/app-luatest/shutdown_test.lua
@@ -273,3 +273,37 @@ g_vinyl.test_shutdown_vinyl_dump_during_snapshot = function(cg)
     fiber.sleep(3)
     test_no_hang_on_shutdown(cg.server)
 end
+
+local g_engine = t.group('engine', {{engine = 'memtx'}, {engine = 'vinyl'}})
+
+g_engine.before_each(function(cg)
+    cg.server = server:new()
+    cg.server:start()
+end)
+
+g_engine.after_each(function(cg)
+    if cg.server ~= nil then
+        cg.server:drop()
+    end
+end)
+
+-- Test we can interrupt loop on secondary index creation.
+g_engine.test_shutdown_secondary_index_build = function(cg)
+    t.tarantool.skip_if_not_debug()
+    cg.server:exec(function(engine)
+        local fiber = require('fiber')
+        box.schema.create_space('test', {engine = engine})
+        box.space.test:create_index('pk')
+        fiber.set_slice(100)
+        box.begin()
+        for i=1,10000 do
+            box.space.test:insert{i, i}
+        end
+        box.commit()
+        box.error.injection.set('ERRINJ_BUILD_INDEX_TIMEOUT', 0.01)
+        fiber.new(function()
+            box.space.test:create_index('sk', {parts = {2}})
+        end)
+    end, {cg.params.engine})
+    test_no_hang_on_shutdown(cg.server)
+end

```

</details>